device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsSessionHandler.java

```diff
@@ -338,110 +338,121 @@ void openLinks()
-                    {
-                        createMethodLinks();
-                        this.explicitInProgressMethodsSubscriptionMessage = transportMessage;
-                        return true; //connection layer doesn't care about this delivery tag
-                    }
-                    else
-                    {
-                        log.debug("Rejecting methods subscription message because that subscription is already in progress");
-                        return false;
-                    }
-                }
-                else //noinspection StatementWithEmptyBody
-                    if (subscriptionType == DEVICE_OPERATION_TWIN_UNSUBSCRIBE_DESIRED_PROPERTIES_REQUEST)
-                {
-                    //TODO: can add logic here to tear down twin links if the user wants to unsubscribe to desired properties
-                }
-                else if (subscriptionType == DEVICE_OPERATION_TWIN_SUBSCRIBE_DESIRED_PROPERTIES_REQUEST)
-                {
-                    if (this.twinSenderLinkOpened && this.twinReceiverLinkOpened)
-                    {
-                        // No need to do anything. Twin links are already opened and desired properties subscription is automatically
-                        // sent once the twin links are opened.
-                        this.amqpsSessionStateCallback.onMessageAcknowledged(message, Accepted.getInstance(), this.getDeviceId());
-                        return true;
-                    }
+        //Check if the message being sent is a subscription change message. If so, open the corresponding links.
+        if (message instanceof IotHubTransportMessage)
+        {
+            IotHubTransportMessage transportMessage = (IotHubTransportMessage) message;
+            DeviceOperations subscriptionType = ((IotHubTransportMessage) message).getDeviceOperationType();
 
-                    if (this.explicitInProgressTwinSubscriptionMessage == null)
-                    {
-                        createTwinLinks();
-                        this.explicitInProgressTwinSubscriptionMessage = transportMessage;
-                        return true;
-                    }
-                    else
-                    {
-                        log.debug("Rejecting twin subscription message because that subscription is already in progress");
-                        return false;
-                    }
-                }
+            if (subscriptionType == DEVICE_OPERATION_METHOD_SUBSCRIBE_REQUEST)
+            {
+                return handleMethodSubscriptionRequest(transportMessage);
+            }
+            //else if (subscriptionType == DEVICE_OPERATION_TWIN_UNSUBSCRIBE_DESIRED_PROPERTIES_REQUEST)
+            //{
+            //    //TODO: can add logic here to tear down twin links if the user wants to unsubscribe to desired properties
+            //}
+            else if (subscriptionType == DEVICE_OPERATION_TWIN_SUBSCRIBE_DESIRED_PROPERTIES_REQUEST)
```

Comment:
As 363 returns, doesn't really need `else`

Reply:
Got rid of the ```else```, not the ```else if``` as we discussed